SimPEG/base/pde_simulation.py

```diff
@@ -1,41 +1,72 @@
 import numpy as np
 import scipy.sparse as sp
-from discretize.utils import Zero
+from discretize.utils import Zero, TensorType
 from ..simulation import BaseSimulation
 from .. import props
 from scipy.constants import mu_0
 
 
 def __inner_mat_mul_op(M, u, v=None, adjoint=False):
     u = np.squeeze(u)
-    if v is not None:
-        if v.ndim > 1:
-            v = np.squeeze(v)
-        if u.ndim > 1:
-            # u has multiple fields
-            if v.ndim == 1:
-                v = v[:, None]
-        else:
+    if sp.issparse(M):
+        if v is not None:
             if v.ndim > 1:
-                u = u[:, None]
-        if v.ndim > 2:
-            u = u[:, None, :]
-        if adjoint:
-            if u.ndim > 1 and u.shape[-1] > 1:
-                return M.T * (u * v).sum(axis=-1)
-            return M.T * (u * v)
-        if u.ndim > 1 and u.shape[1] > 1:
-            return np.squeeze(u[:, None, :] * (M * v)[:, :, None])
-        return u * (M * v)
+                v = np.squeeze(v)
+            if u.ndim > 1:
+                # u has multiple fields
+                if v.ndim == 1:
+                    v = v[:, None]
+            else:
+                if v.ndim > 1:
+                    u = u[:, None]
+            if v.ndim > 2:
+                u = u[:, None, :]
+            if adjoint:
+                if u.ndim > 1 and u.shape[-1] > 1:
+                    return M.T * (u * v).sum(axis=-1)
+                return M.T * (u * v)
+            if u.ndim > 1 and u.shape[1] > 1:
+                return np.squeeze(u[:, None, :] * (M * v)[:, :, None])
+            return u * (M * v)
+        else:
+            if u.ndim > 1:
+                UM = sp.vstack([sp.diags(u[:, i]) @ M for i in range(u.shape[1])])
+            else:
+                U = sp.diags(u, format="csr")
+                UM = U @ M
+            if adjoint:
+                return UM.T
+            return UM
     else:
+        # assume it was a tuple of M_func, prop_deriv
```

Comment:
would it be clearer if in the first check we looked to see if M is a tuple? (or list) here and then error if we get something unexpected?

Reply:
I've added a more explicit check and an appropriate error, instead of just an assumption. This should hopefully only ever be an issue if a developer does something unexpected with new code.